[quant][fx][graphmode] Support standalone module in _convert_do_not_use

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70152
* #70287
* #70286
* __->__ #70151

Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

jerryzh168 committed:
[quant][fx][graphmode] Support standalone module in _convert_do_not_use

Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

supriyar commented on torch/ao/quantization/fx/_convert_do_not_use.py:
```diff
@@ -190,12 +202,55 @@ def replace_observer_with_quantize_dequantize_node(graph: Graph, node: Node, mod
+                for idx in range(len(args)):
+                    if idx in sm_input_quantized_idxs:
+                        arg = args[idx]
+                        if arg.op == "call_method" and arg.target == "dequantize":
```
If the input index is already quantized, why do we expect to have a `dequantize` here for the input index?

jerryzh168 replied:
oh, this is because the new convert flow will output reference pattern by default, and all operators in this flow are assumed to take fp32 inputs, so we'd need to dequantize the input here so that it can be consumed by the following operator

supriyar replied:
I see, so the `quantize_node` on line 227 is not really a `quantize_per_tensor` node, but a node with quantized value?

jerryzh168 replied:
right it's a bit confusing...it actually means both, e.g.
quantize_node means the following node:
```
quantized = call_function[target=torch.quantize_per_tensor](input)
```
but we can also use quantize_node as an argument (value):
```
dequantized = call_method[target="dequantize"](quantized)
```
it comes from the fact that Node in fx means both the operator and the output value, there is no separate value concept in fx graph I think. This also causes some other confusion in our code base.
cc @vkuzo @jamesr66a for any additional comments

supriyar reviewed (approved):
looks good

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Support standalone module in _convert_do_not_use"


Summary:
this supports converting an observed standalone module to quantized standalone module
in the new convert flow (convert observers to quant-dequant operators)

Test Plan:
```
python test/test_quant_trt.py TestConvertFxDoNotUse
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205163](https://our.internmc.facebook.com/intern/diff/D33205163)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205163).
